internal/otelcollector/config/tracegateway/config_builder.go

```diff
@@ -32,94 +32,284 @@ type BuildOptions struct {
-		},
-		Receivers:  receiversConfig(),
-		Processors: processorsConfig(opts),
-		Exporters:  make(Exporters),
+// buildComponentFunc defines a function type for building components in the telemetry configuration
+type buildComponentFunc = func(ctx context.Context, tp *telemetryv1alpha1.TracePipeline) error
+
+// componentConfigFunc creates the configuration for a component (receiver or processor)
+type componentConfigFunc func(tp *telemetryv1alpha1.TracePipeline) any
+
+// exporterComponentConfigFunc creates the configuration for an exporter component
+// creating exporters is different from receivers and processors, as it makes an API server call to resolve the reference secrets
+// and returns the configuration along with environment variables needed for the exporter
+type exporterComponentConfigFunc func(ctx context.Context, tp *telemetryv1alpha1.TracePipeline) (any, common.EnvVars, error)
+
+// componentIDFunc determines the ID of a component
+type componentIDFunc func(*telemetryv1alpha1.TracePipeline) string
+
+// staticComponentID returns a ComponentIDFunc that always returns the same component ID independent of the TracePipeline
+func staticComponentID(componentID string) componentIDFunc {
+	return func(*telemetryv1alpha1.TracePipeline) string {
+		return componentID
 	}
 }
 
-func receiversConfig() Receivers {
-	return Receivers{
-		OTLP: common.OTLPReceiver{
-			Protocols: common.ReceiverProtocols{
-				HTTP: common.Endpoint{
-					Endpoint: fmt.Sprintf("${%s}:%d", common.EnvVarCurrentPodIP, ports.OTLPHTTP),
-				},
-				GRPC: common.Endpoint{
-					Endpoint: fmt.Sprintf("${%s}:%d", common.EnvVarCurrentPodIP, ports.OTLPGRPC),
-				},
-			},
-		},
+func (b *Builder) addServicePipeline(ctx context.Context, pipeline *telemetryv1alpha1.TracePipeline, fs ...buildComponentFunc) error {
+	for _, f := range fs {
+		if err := f(ctx, pipeline); err != nil {
+			return fmt.Errorf("failed to add component: %w", err)
+		}
 	}
+
+	return nil
 }
 
-// addComponentsForTracePipeline enriches a Config (exporters, processors, etc.) with components for a given telemetryv1alpha1.TracePipeline.
-func (b *Builder) addComponentsForTracePipeline(ctx context.Context, pipeline *telemetryv1alpha1.TracePipeline, queueSize int) error {
-	b.addUserDefinedTransformProcessor(pipeline)
-	return b.addOTLPExporter(ctx, pipeline, queueSize)
+// withReceiver creates a decorator for adding receivers
+func (b *Builder) withReceiver(componentIDFunc componentIDFunc, configFunc componentConfigFunc) buildComponentFunc {
+	return func(ctx context.Context, tp *telemetryv1alpha1.TracePipeline) error {
+		componentID := componentIDFunc(tp)
+
+		config := configFunc(tp)
+		if config == nil {
+			// If no config is provided, skip adding the receiver
+			return nil
+		}
+
+		if _, found := b.config.Receivers[componentID]; !found {
+			b.config.Receivers[componentID] = config
+		}
+
+		pipelineID := formatTraceServicePipelineID(tp)
+		pipeline := b.config.Service.Pipelines[pipelineID]
```

Comment:
where to we initially add add the `pipelineID` to the map b.config.Service.Pipelines?

Reply:
It works because of this behavior of Go maps https://go.dev/blog/maps#exploiting-zero-values. But I agree, an explicit call to populate the map will be better.